openmdao/solvers/nonlinear/broyden.py

```diff
@@ -460,19 +515,85 @@ def get_states(self):
         ndarray
             Array containing values of states.
         """
+        system = self._system
         outputs = self._system._outputs
+        states = self.options['state_vars']
 
-        if self._full_inverse:
+        if system.comm.size > 1:
+            states = None if not states else states
+
+            out_vec = outputs._data
+            mpi_typ = MPI.C_DOUBLE_COMPLEX if system.under_complex_step else MPI.DOUBLE
+
+            _, _, local2owned_inds, noncontig_dist_inds = \
+                system._get_nodup_out_ranges(var_list=states)
+
+            # gather the 'owned' parts of out_vec from each process
+            xm = np.empty(self.size, dtype=out_vec.dtype)
+            system.comm.Gatherv((out_vec[local2owned_inds], local2owned_inds.size, mpi_typ),
+                                (xm, self._sendcounts, mpi_typ),
+                                root=0)
+
+            # Convert full_b to the same ordering that the matrix expects, where
+            # dist vars are contiguous and other vars appear in 'execution' order.
+            xm = xm[noncontig_dist_inds]
+
+        elif self._full_inverse:
             xm = outputs._data.copy()
+
         else:
-            states = self.options['state_vars']
             xm = self.xm.copy()
             for name in states:
                 i, j = self._idx[name]
                 xm[i:j] = outputs[name]
 
         return xm
 
+    def get_residuals(self):
+        """
+        Return a vector containing the residuals of the states specified in options.
+
+        Returns
+        -------
+        ndarray
+            Array containing values of residuals.
+        """
+        system = self._system
+        residuals = system._residuals
+        states = self.options['state_vars']
+
+        if system.comm.size > 1:
+            states = None if not states else states
+
+            mpi_typ = MPI.C_DOUBLE_COMPLEX if system.under_complex_step else MPI.DOUBLE
+            res_vec = residuals._data
+
+            _, _, local2owned_inds, noncontig_dist_inds = \
+                system._get_nodup_out_ranges(var_list=states)
+
+            # gather the 'owned' parts of res_vec from each process
+            fxm = np.empty(self.size, dtype=res_vec.dtype)
+            system.comm.Gatherv((res_vec[local2owned_inds], local2owned_inds.size, mpi_typ),
+                                (fxm, self._sendcounts, mpi_typ),
+                                root=0)
+
+            # Convert full_b to the same ordering that the matrix expects, where
+            # dist vars are contiguous and other vars appear in 'execution' order.
+            fxm = fxm[noncontig_dist_inds]
+
+            self.fxm = fxm
+
+        elif self._full_inverse:
+            fxm = self.fxm = residuals._data.copy()
+
+        else:
+            fxm = self.fxm
+            for name in states:
+                i, j = self._idx[name]
+                fxm[i:j] = residuals[name]
+
+        return fxm
```

Comment:
it looks like the logic is identical to `get_states`...  did you consider something like adding a boolean `residuals` arg to select the appropriate vector rather than duplicating the code?

Reply:
Yeah, got this to work.